harp/reader.py

```diff
@@ -51,13 +51,21 @@ class DeviceReader:
     def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
         self.device = device
         self.registers = registers
+        self._register_map = {self.registers[key].register.address: key for key in self.registers.keys()}
 
     def __dir__(self) -> Iterable[str]:
         return self.registers.keys()
 
     def __getattr__(self, __name: str) -> RegisterReader:
         return self.registers[__name]
 
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            return self.registers[self._register_map[__key]]
+        elif isinstance(__key, str):
+            return self.registers[__key]
+        else:
+            raise TypeError(f"key must be int or str, not {type(__key)}")
```

Comment:
I think the design would be cleaner if the type polymorphism was provided at the level of the `registers` attribute rather than making the `DeviceReader` class itself a collection. The `__dir__` and `__getattr__` methods were implemented only to provide a convenient "dot" notation into the object structure, but for dynamic accesses by key I was considering using the public `registers` attribute instead.